c++/src/kj/compat/http.c++

```diff
@@ -6368,61 +6489,54 @@ public:
-
-      // This operates optimistically. In order to support pipelining, we have to
-      // accept the tunnel early, even if we're not yet certain that the tunnel can
-      // actually be established. If the response.status promise rejects or returns
-      // a failure, then we will interrupt the pipelines we establish below.
-      auto io = kj::mv(response.connection);
-      HttpHeaderTable headerTable {};
-      HttpHeaders responseHeaders(headerTable);
-      auto io2 = tunnel.accept(200, "OK"_kj, responseHeaders);
-
-      auto promises = kj::heapArrayBuilder<kj::Promise<void>>(2);
-      promises.add(io->pumpTo(*io2).then([&io2 = *io2](uint64_t size) {
+    auto request = client.connect(host, headers);
+
+    // This operates optimistically. In order to support pipelining, we connect the
+    // input and outputs streams immediately, even if we're not yet certain that the
+    // tunnel can actually be established.
+    auto promises = kj::heapArrayBuilder<kj::Promise<void>>(2);
+    auto io1 = kj::mv(request.connection);
+    auto io2 = kj::mv(connection);
+
+    // Here, we immediately set up the outbound pipe... that is, piping from
+    // the passed in connection to the newly created request.connection.
+    // However, we do not set up the pipeline in the other direction until
+    // we have received the status from the connection.
+
+    promises.add(io2->pumpTo(*io1).then([&io1 = *io1](uint64_t size) {
+      io1.shutdownWrite();
+    }).eagerlyEvaluate(nullptr));
+
+    promises.add(request.status.then(
+        [&tunnel,&io1=*io1,&io2=*io2](HttpClient::ConnectRequest::Status status) {
+      if (status.statusCode >= 200 && status.statusCode < 300) {
+        // In the success case, we pipeline from the request.connection to the
+        // passed in connection, fully establishing the bidirectional tunnel.
+        tunnel.accept(status.statusCode, status.statusText, *status.headers);
+        return io1.pumpTo(io2).then([&io2](uint64_t size) {
+          io2.shutdownWrite();
+        });
+      } else {
+        // However, if the connect request is rejected, we want to shutdown the tunnel
+        // and pipeline the request.connection to the AsyncOutputStream returned by
+        // reject.
         io2.shutdownWrite();
-      }).eagerlyEvaluate(nullptr));
-      promises.add(io2->pumpTo(*io).then([&io = *io](uint64_t size) {
-        io.shutdownWrite();
-      }).eagerlyEvaluate(nullptr));
-
-      return maybeCanceler->wrap(
-          kj::joinPromises(promises.finish()).attach(kj::mv(io), kj::mv(io2)))
-              .attach(kj::mv(maybeCanceler));
-    });
+        auto out = tunnel.reject(status.statusCode, status.statusText, *status.headers);
+        return io1.pumpTo(*out).then([](uint64_t) -> kj::Promise<void> {
+          return kj::READY_NOW;
+        }).attach(kj::mv(out));
+      }
+    }).eagerlyEvaluate(nullptr));
+
+    return kj::joinPromises(promises.finish()).attach(kj::mv(io1), kj::mv(io2));
```

Comment:
I'm a bit concerned that if the pump in one direction throws an error, it'll get held in limbo waiting for the other direction to resolve. This seems like it could easily hang.

However, I notice the handling of WebSockets appears to suffer from the same problem. Maybe it's a bug we should fix for both in a later PR? Maybe add a `// TODO(bug)` for now?

Reply:
Yeah, I actually meant to add a comment here asking about that, I'm glad you mentioned it because I wasn't certain.

Reply:
I'm not entirely clear on the right answer either.

Consider the case where the client has established a connection to the server, and the down-direction pump (from server to client) throws an exception. What should happen?

* We could immediately cancel the up-direction pump as well, and fail out everything.

* We could let the up direction continue, but try to make sure that any `read()`s issued by the client throw an exception indicating the down direction is broken.

The problem right now is we don't do _either_. The `read()` will just hang.

I would say the second option seems better, but I don't think we have any way to implement it with the current stream APIs. What you want here is an `abortWrite()` call, which causes `read()` on the other end of the pipe to throw. But `AsyncIoStream` only has `abortRead()` (causes `write()` on the other end to throw) and `shutdownWrite()` (which signals a clean EOF to the reader -- not what we want!).

Perhaps in addition to adding an `end()` method I really need to add an `abortWrite()` method. But as of now, we don't have that.

So perhaps the only option now is to cancel the opposite direction when one direction fails. I guess the easiest way to do that is using a `kj::Canceler`. It's an annoying amount of bookkeeping, though. It would be nice to fix this properly by adding a `kj::joinPromisesFailfast` to supplement (or perhaps replace) `kj::joinPromises`, and then use that here. (This new function would avoid the lazy evaluation problem, too.)

Reply:
But yeah I'm also OK with `TODO(bug)` for now.